oeps/oep-0017-feature-toggles.rst

```diff
@@ -0,0 +1,809 @@
+expose a new feature on a specific grand opening date. However, for a confident
+unveiling, this use case should be used in consideration with Ops_ and/or `Beta
+Testing`_ scenarios.
+
+Ops
+~~~
+
+Dynamically controlling feature toggles, without needing to re-deploy an application,
+comes in very handy when considering the operational requirements for uptime metrics.
+
+This use case is usually driven by the engineering team.
+
+Use Case 3: Ops - Monitored Rollout
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. _Ops - Monitored Rollout: `Use Case 3: Ops - Monitored Rollout`_
+
+As teams balance the needs for rapid agile development while continuously deploying
+to a large-scale system with 99.99% uptime requirements, they need the ability to
+test new changes in production while having the ability to revert quickly. That is,
+moving rapidly and taking risks can decrease Mean Time to Failure (MTTF), which needs
+to be counterbalanced with the ability to reduce `Mean Time to Recovery (MTTR)`_.
+
+When a team is concerned about potential performance or scalability issues with an
+upcoming change, gating the change behind a toggle allows the team to:
+
+* control when the change is enabled so they can monitor it in production at their
+  own time, independent of the deployment cycle
+
+* quickly disable the change in case of unexpected issues in production
+
+* gradually rollout the change (`canary release`_), starting with a small percentage
+  of random users, detecting regressions, addressing any issues that arise, before
+  enabling for everyone.
+
+Once the team is confident about their change and the change is released to all users,
+they would safely remove the gating toggle.
+
+.. _Mean Time to Recovery (MTTR): https://www.thoughtworks.com/radar/techniques/focus-on-mean-time-to-recovery
+.. _canary release: https://martinfowler.com/bliki/CanaryRelease.html
+
+Use Case 4: Ops - Graceful Degradation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. _Ops - Graceful Degradation: `Use Case 4: Ops - Graceful Degradation`_
+
+In certain cases, the development team (in consultation with the operations team)
+may choose to extend the lifetime of an Ops toggle in the codebase even after
+releasing its gated feature. A small number of such long-lived Ops "kill switches"
+provide operators dynamic controls to gracefully degrade the system under high load.
+Operators can use these circuit-breaker capabilities either preemptively in the
+anticipation of a high-demand event or in response to taming an unanticipated high
+load or attack.
+
+Typically, long-lived Ops toggles are useful for gating non-critical features that
+are very expensive on system resources. However, the long-term costs of maintaining
+the added complexity in the code should be measured against the benefits of
+operationally degrading the service when needed.
+
+Use Case 5: Beta Testing
+~~~~~~~~~~~~~~~~~~~~~~~~
+.. _Beta Testing: `Use Case 5: Beta Testing`_
```

Comment:
How does this differ from an A/B test?  Isn't this an A/B test where A is the current system and B is the new thing we thing we're trying out?  Is it that our current A/B framework doesn't allow for targetting in the ways you mentionned below or some other reason?